aten/src/ATen/native/TensorFactories.cpp

```diff
@@ -415,6 +415,18 @@ Tensor ones_like(
 // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ scalar_tensor ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
 Tensor scalar_tensor(Scalar s, const TensorOptions& options) {
+  if (options.device() == at::kCPU) {
+    // This is a fast track to skip device dispatch for making scalar tensor on CPU.
+    // See https://github.com/pytorch/pytorch/pull/29915 for more detailed perf
+    // difference.
+    // In the future when we remove the overhead of device dispatch, we'll happily
+    // revert this to following:
+    //   auto result = at::empty({}, options);
+    at::AutoNonVariableTypeMode non_var_type_mode(true);
```

Comment:
@ezyang @ngimel I did a minor change in the latest commit since [`empty_cpu` requires VariableTypeId to be excluded](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/TensorFactories.cpp#L102).
Currently I'm using the `AutoNonVariableTypeMode` here to turn it off and then directly call into `empty_cpu`. Alternatively I can also use `auto result = at::empty({}, options)` but there's a small difference in perf (of the same script in description). 
```
## with AutoNonVariableTypeMode + empty_cpu
61.3 µs ± 165 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
## with at::empty
63.1 µs ± 173 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```
Since the total time before this PR was ~80us, the 2us difference here doesn't sound too appealing. But I'm putting it here, just checking whether we want this 2us when we know for sure something is definitely on CPU.

Reply:
Use of AutoNonVariableTypeMode here is fine. (It's a bit of black magic inserting these, but that's what we're doing right now.)